### PR TITLE
[DDCNL-7800]tai-frontend re-write journey cache: Benefits

### PR DESCRIPTION
**Pull request by vfryer12**
- [x] Unit tests passing
- [x] Coverage reached
- [x] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

**Review by grantrobson — COMMENTED** [review]
Added some comments

File: app/controllers/benefits/CompanyBenefitController.scala [hunk 1]
```diff
@@ -16,84 +16,89 @@
...
   authenticate: AuthJourney,
   mcc: MessagesControllerComponents,
   updateOrRemoveCompanyBenefitDecision: UpdateOrRemoveCompanyBenefitDecisionView,
-  implicit val errorPagesHandler: ErrorPagesHandler
+  journeyCacheNewRepository: JourneyCacheNewRepository,
+  errorPagesHandler: ErrorPagesHandler
 )(implicit ec: ExecutionContext)
     extends TaiBaseController(mcc) with Logging {
 
+  private val journeyStartRedirection = Redirect(controllers.routes.TaxAccountSummaryController.onPageLoad().url)
```
[the remark this answers is not in the document]
Comment (vfryer12) [reply app/controllers/benefits/CompanyBenefitController.scala]: updated

File: app/controllers/benefits/CompanyBenefitController.scala [hunk 2]
```diff
@@ -111,27 +116,44 @@ class CompanyBenefitController @Inject() (
             .auditAndRedirectService(TaiConstants.CompanyBenefitsIform)
         )
       case _ =>
-        logger.error(s"Bad Option provided in submitDecision form: $decision")
         failureRoute
     }
 
-  def submitDecision: Action[AnyContent] = authenticate.authWithValidatePerson.async { implicit request =>
+  private def cacheDecision(
```
[the remark this answers is not in the document]
Comment (vfryer12) [reply app/controllers/benefits/CompanyBenefitController.scala]: updated

File: app/controllers/benefits/CompanyBenefitsSummaryController.scala [hunk 3]
```diff
@@ -38,67 +36,76 @@ import scala.concurrent.{ExecutionContext, Future}
...
+          )
+          val result = if (!incomeDetailsViewModel.isUpdateInProgress) {
+            journeyCacheNewRepository.clear(request.userAnswers.sessionId, nino.nino).map(_ => (): Unit)
+          } else {
+            Future.successful((): Unit)
+          }
+          val person: Future[Person] = personService.personDetailsFuture(nino)
+          person.flatMap { person =>
+            implicit val messages: Messages = request2Messages(request)
+            implicit val authRequest: AuthenticatedRequest[AnyContent] =
```
[the remark this answers is not in the document]
Comment (vfryer12) [reply app/controllers/benefits/CompanyBenefitsSummaryController.scala]: updated

File: app/controllers/benefits/RemoveCompanyBenefitController.scala [hunk 4]
```diff
@@ -56,141 +57,170 @@ class RemoveCompanyBenefitController @Inject() (
...
-          currentCache(EndCompanyBenefitConstants.EmploymentNameKey)
+          benefitName,
+          employmentName
         )
       )
     }
   }
 
-  def submitStopDate: Action[AnyContent] = authenticate.authWithValidatePerson.async { implicit request =>
+  // scalastyle:off method.length
```
[the remark this answers is not in the document]
Comment (vfryer12) [reply app/controllers/benefits/RemoveCompanyBenefitController.scala]: updated

File: app/controllers/benefits/RemoveCompanyBenefitController.scala [hunk 5]
```diff
@@ -56,141 +57,170 @@ class RemoveCompanyBenefitController @Inject() (
...
+
+                Json.toJson(updatedData) match {
+                  case jsObject: JsObject => journeyCacheNewRepository.set(userAnswers.copy(data = jsObject))
+                  case _                  => Future.failed(new Exception("Updated data is not a JSON object"))
                 }
-            }
+              }
+            } yield Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.telephoneNumber())
+          } else {
+            // OnOrAfterTaxYearEnd
```
[the remark this answers is not in the document]
Comment (vfryer12) [reply app/controllers/benefits/RemoveCompanyBenefitController.scala]: updated

File: app/controllers/benefits/RemoveCompanyBenefitController.scala [hunk 6]
```diff
@@ -56,141 +57,170 @@ class RemoveCompanyBenefitController @Inject() (
...
+            )
+          )
+        },
+        { date =>
+          val dateString = date.toString
+          if (date isBefore taxYear.start) {
+            for {
+              _ <- journeyCacheNewRepository.clear(userAnswers.sessionId, user.nino.nino)
+              _ <- {
+                val current = userAnswers.data.as[Map[String, JsValue]]
```
[the remark this answers is not in the document]
Comment (vfryer12) [reply app/controllers/benefits/RemoveCompanyBenefitController.scala]: updated

**Review by grantrobson — APPROVED** [review]

**Pull request merged**